### ISSUE TITLE
Research: Benchmark Czech vs English semantic search accuracy

### ISSUE THREAD

**Issue by Olbrasoft** [Olbrasoft/HandbookSearch#19]
## Overview

Test and measure whether Czech embeddings actually improve semantic search accuracy for Czech queries compared to cross-lingual search (Czech query → English documents).

## Background

Recent research on multilingual embeddings (2024-2025) shows:

### Key Findings from Research

**1. Language Affinity Bias**
> Even when limiting inputs to languages officially supported by an embedding model, the outputs will still yield similarity scores that reflect to some degree language affinity, with content in the same language or in languages that cooccur more frequently on the web having higher similarity scores. ([GDELT Project](https://blog.gdeltproject.org/multilingual-embedding-for-llm-external-memory-semantic-search-universal-sentence-encoder-family-labse-vertex-ai-embeddings-for-text/))

**Solution**: Either translate all content to unified language OR have separate embeddings per language.

**2. Monolingual vs Multilingual Performance**
> Monolingual LLMs give better results for verse classification and matching verse retrieval compared to multilingual models. ([MDPI Research](https://www.mdpi.com/2078-2489/16/2/84))

**3. BGE-M3 Acknowledgement**
> BGE-M3 acknowledges that its ability in monolingual embedding retrieval may not surpass models specifically designed for single languages. ([Modal Blog](https://modal.com/blog/mteb-leaderboard-article))

**4. Czech-Specific Research**
> Small Czech embedding models (15-20M params) matched or exceeded BASE-sized multilingual models in semantic similarity tasks. ([arXiv](https://arxiv.org/html/2311.13921))

**5. Cross-Lingual Accuracy**
> Slovak-Czech cross-lingual accuracy achieves ~70%, while same-language performance is significantly higher. ([Seznam GitHub](https://github.com/seznam/czech-semantic-embedding-models))

### Hypothesis

**Czech query → Czech documents** will have better semantic similarity scores than **Czech query → English documents** using nomic-embed-text multilingual embeddings.

### Test Queries (Czech)

Create 10-15 test queries in Czech covering different topics from engineering-handbook:

```
1. "jak nastavit CI/CD pipeline pro .NET projekt"
2. "jak používat git branches a pull requesty"
3. "SOLID principy v C#"
4. "jak testovat ASP.NET Core aplikaci"
5. "jak spravovat tajné klíče v projektu"
6. "návrh čisté architektury"
7. "deployment do produkce"
8. "code review best practices"
9. "jak psát unit testy s xUnit"
10. "PostgreSQL migrace s Entity Framework"
```

### Ground Truth

For each query, manually identify 2-3 **most relevant** documents from engineering-handbook (English version).

Example:
- Query: "jak nastavit CI/CD pipeline"
- Expected top results:
  1. `development-guidelines/dotnet/continuous-integration/index-continuous-integration.md`
  2. `development-guidelines/dotnet/continuous-deployment/index-continuous-deployment.md`

### Test Scenarios

**Scenario A: Cross-Lingual Search (Current)**
- Query language: Czech
- Document language: English
- Embedding: nomic-embed-text (multilingual)

**Scenario B: Same-Language Search (After Implementation)**
- Query language: Czech
- Document language: Czech (translated)
- Embedding: nomic-embed-text (multilingual)

### Metrics

For each scenario, measure:

1. **Precision@K**
   - Precision@3: How many of top 3 results are relevant?
   - Precision@5: How many of top 5 results are relevant?

2. **Mean Reciprocal Rank (MRR)**
   - Position of first relevant result
   - Formula: `MRR = 1 / rank_of_first_relevant`

3. **NDCG@K (Normalized Discounted Cumulative Gain)**
   - Quality of ranking with position weighting
   - Higher NDCG = better ranking

4. **Average Cosine Distance**
   - Lower distance = better semantic match
   - Compare average distance to ground truth documents

### Success Criteria

Czech embeddings are considered **beneficial** if:
- ✅ Precision@3 improves by ≥10%
- ✅ MRR improves by ≥15%
- ✅ Average cosine distance decreases by ≥20%

### Phase 1: Baseline Testing (Cross-Lingual)

1. Deploy current HandbookSearch with English documents
2. Run 10-15 Czech test queries
3. Record top-5 results for each query
4. Calculate baseline metrics

### Phase 2: Czech Embeddings Testing (Same-Language)

1. Complete issues #15-#17 (translation + import)
2. Run same test queries against Czech documents
3. Record top-5 results
4. Calculate metrics

### Phase 3: Comparison

1. Compare metrics side-by-side
2. Analyze improvement (or degradation)
3. Document findings

### Phase 4: Real-World Validation

Test with actual user prompts:
- Collect 20-30 real Czech prompts from usage logs
- Measure which results users find more helpful
- A/B test if possible

## Test Tool

Create simple benchmark CLI tool:

```bash
cd ~/Olbrasoft/HandbookSearch/src/HandbookSearch.Benchmark

# Run baseline test
dotnet run -- benchmark \
  --queries-file test-queries-cs.json \
  --ground-truth ground-truth.json \
  --language en \
  --output baseline-results.json

# Run Czech test
dotnet run -- benchmark \
  --queries-file test-queries-cs.json \
  --ground-truth ground-truth.json \
  --language cs \
  --output czech-results.json

# Compare
dotnet run -- compare \
  --baseline baseline-results.json \
  --experiment czech-results.json \
  --output comparison-report.md
```

### Test Data Format

**test-queries-cs.json**:
```json
[
  {
    "id": 1,
    "query": "jak nastavit CI/CD pipeline pro .NET projekt",
    "topic": "continuous-integration"
  },
  ...
]
```

**ground-truth.json**:
```json
[
  {
    "queryId": 1,
    "relevantDocs": [
      "development-guidelines/dotnet/continuous-integration/index-continuous-integration.md",
      "development-guidelines/dotnet/continuous-deployment/index-continuous-deployment.md"
    ],
    "rankings": {
      "development-guidelines/dotnet/continuous-integration/index-continuous-integration.md": 1,
      "development-guidelines/dotnet/continuous-deployment/index-continuous-deployment.md": 2
    }
  },
  ...
]
```

### If Czech Embeddings WIN

- **Result**: Proceed with full implementation
- **Action**: Import all Czech translations (#17)
- **Next**: Integrate with Claude Code workflow (#19)

### If Cross-Lingual WINS or TIE

- **Result**: Re-evaluate translation strategy
- **Analysis**: Why didn't it help?
  - nomic-embed-text multilingual quality too good?
  - Translation quality issues?
  - Test queries not representative?
- **Decision**: May not need Czech translations for search

### If Results are MIXED

- **Result**: Conditional implementation
- **Analysis**: Which query types benefit from Czech?
- **Decision**: Selective translation (only certain topics)

## Dependencies

- **Depends on**: #15 (Database schema)
- **Depends on**: #16 (Translation program)
- **Depends on**: #17 (Czech embeddings import)
- **Blocks**: #19 (Workflow integration - pending results)

## References

**Research Papers & Articles**:
- [Multilingual Embedding for Semantic Search (GDELT)](https://blog.gdeltproject.org/multilingual-embedding-for-llm-external-memory-semantic-search-universal-sentence-encoder-family-labse-vertex-ai-embeddings-for-text/)
- [Fine-Tuning on Monolingual and Multilingual Models (MDPI)](https://www.mdpi.com/2078-2489/16/2/84)
- [Czech Semantic Embedding Models (arXiv)](https://arxiv.org/html/2311.13921)
- [Czech Semantic Embedding Models (GitHub)](https://github.com/seznam/czech-semantic-embedding-models)
- [Top Embedding Models MTEB Leaderboard (Modal)](https://modal.com/blog/mteb-leaderboard-article)
- [Nomic Embed Text V2 (Nomic Blog)](https://www.nomic.ai/blog/posts/nomic-embed-text-v2)

**nomic-embed-text Specifics**:
- Multilingual support: ~100 languages
- Training data: 1.6B pairs from mC4 and multilingual CC News
- MIRACL benchmark: SOTA multilingual performance
- Technical paper: [arXiv:2502.07972](https://arxiv.org/pdf/2502.07972)

## Acceptance Criteria

- [ ] Test queries defined (10-15 Czech queries)
- [ ] Ground truth documented
- [ ] Baseline metrics collected (cross-lingual)
- [ ] Czech embeddings metrics collected (same-language)
- [ ] Comparison report generated
- [ ] Decision documented: Proceed / Re-evaluate / Conditional
- [ ] Findings summarized in issue comment

## Timeline

- **Phase 1** (Baseline): 2-3 hours (after current English import)
- **Phase 2** (Czech test): 2-3 hours (after issues #15-17 complete)
- **Phase 3** (Analysis): 1-2 hours
- **Total**: ~1 day of work

## Notes

This research issue should be completed **BEFORE** implementing workflow integration (#19). We need empirical evidence that Czech embeddings actually improve search quality.

If results show minimal improvement (<10%), we may reconsider the entire translation approach.
